ingestion-beam/src/main/java/com/mozilla/telemetry/contextualservices/AggregateImpressions.java

```diff
@@ -0,0 +1,113 @@
+      ParsedReportingUrl.PARAM_FORM_FACTOR, ParsedReportingUrl.PARAM_OS_FAMILY,
+      ParsedReportingUrl.PARAM_ID);
+
+  private final ValueProvider<String> aggregationWindowDuration;
+
+  public AggregateImpressions(ValueProvider<String> aggregationWindowDuration) {
+    this.aggregationWindowDuration = aggregationWindowDuration;
+  }
+
+  public static AggregateImpressions of(ValueProvider<String> getAggregationWindowSize) {
+    return new AggregateImpressions(getAggregationWindowSize);
+  }
+
+  @Override
+  public PCollection<PubsubMessage> expand(PCollection<PubsubMessage> messages) {
+    return messages //
+        // Add reporting url as key, pubsub message as value
+        .apply(WithKeys.of((SerializableFunction<PubsubMessage, String>) //
+        AggregateImpressions::getAggregationKey)) //
+        .setCoder(KvCoder.of(StringUtf8Coder.of(), PubsubMessageWithAttributesCoder.of())) //
+        // Set timestamp to current time
+        .apply(WithTimestamps.of(message -> new Instant())) //
+        // Group impressions into timed windows
+        // TODO: Use valueprovider
+        .apply("IntervalWindow", Window.into(FixedWindows.of(Time.parseDuration("10m")))) //
+        .apply(Count.perKey()) //
+        // Create aggregated url by adding impression count as query parameter
+        .apply(ParDo.of(new BuildAggregateUrl())) //
+        .apply("Return to GlobalWindow", Window.into(new GlobalWindows()));
+  }
+
+  @VisibleForTesting
+  static String getAggregationKey(PubsubMessage message) {
+    message = PubsubConstraints.ensureNonNull(message);
+
+    String reportingUrl = message.getAttribute(Attribute.REPORTING_URL);
+
+    ParsedReportingUrl urlParser = new ParsedReportingUrl(reportingUrl);
+
+    // Rebuild url filtering out unneeded parameters
+    ParsedReportingUrl aggregationUrl = new ParsedReportingUrl(urlParser.getBaseUrl());
+    for (String name : aggregationFields) {
+      aggregationUrl.addQueryParam(name, urlParser.getQueryParam(name));
+    }
+
+    return aggregationUrl.toString();
+  }
+
+  @VisibleForTesting
+  static class BuildAggregateUrl extends DoFn<KV<String, Long>, PubsubMessage> {
+
+    @ProcessElement
+    public void processElement(@Element KV<String, Long> input, OutputReceiver<PubsubMessage> out,
+        IntervalWindow window) {
+      ParsedReportingUrl urlParser = new ParsedReportingUrl(input.getKey());
+
+      long impressionCount = input.getValue();
+      long windowStart = window.start().getMillis();
+      long windowEnd = window.end().getMillis();
+
+      urlParser.addQueryParam(ParsedReportingUrl.PARAM_IMPRESSIONS, Long.toString(impressionCount));
+      urlParser.addQueryParam(ParsedReportingUrl.PARAM_TIMESTAMP_BEGIN,
+          Long.toString(windowStart / 1000));
+      urlParser.addQueryParam(ParsedReportingUrl.PARAM_TIMESTAMP_END,
+          Long.toString(windowEnd / 1000));
+
+      Map<String, String> attributeMap = ImmutableMap.of(Attribute.REPORTING_URL,
```

Comment:
submission_timestamp is added for the output written to error output.  Payload and attributes are empty otherwise